vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -594,6 +611,16 @@ func resourceVSphereVirtualMachineRead(d *schema.ResourceData, meta interface{})
 		d.Set("power_state", "suspended")
 	}
 
+	// Check for the presence of a virtual Trusted Platform Module (TPM) device.
+	var isVTPMPresent bool
+	for _, dev := range vprops.Config.Hardware.Device {
+		if _, ok := dev.(*types.VirtualTPM); ok {
+			isVTPMPresent = true
+			break
+		}
+	}
+	_ = d.Set("vtpm_present", isVTPMPresent)
+
```

Comment:
```suggestion
	// Set the virtual Trusted Platform Module device for the virtual machine.
	var isVTPMPresent bool
	for _, dev := range vprops.Config.Hardware.Device {
		if _, ok := dev.(*types.VirtualTPM); ok {
			isVTPMPresent = true
			break
		}
	}

	if isVTPMPresent {
		firmware := vprops.Config.Firmware
		if isVTPMPresent && firmware != string(types.GuestOsDescriptorFirmwareTypeEfi) {
			return errors.New("vtpm requires firmware to be set to efi")
		}
	}

	_ = d.Set("vtpm_present", isVTPMPresent)

```

This should check that the firmware is set to using EFI.